larray/io/excel.py

```diff
@@ -321,10 +321,20 @@ def __setitem__(self, key, value):
         @property
         def shape(self):
             # include top-left empty rows/columns
-            # XXX: is there an exposed xlwings API for this? expand maybe?
+            # see https://www.excelcampus.com/vba/find-last-row-column-cell/#endcode
+            from xlwings.constants import Direction as xldir
             used = self.xw_sheet.api.UsedRange
-            return (used.Row + used.Rows.Count - 1,
-                    used.Column + used.Columns.Count - 1)
+            last_row = max(used.Cells(used.Rows.Count + 1, col).End(xldir.xlUp).Row
```

Comment:
My fear is that this would be too slow for large sheets. Please test this. If it turns out reasonably fast, this is ok. Otherwise, I would use something like the code below (untested!!!) which should be potentially much faster:

```python
last_cell = None
row = used.Row + used.Rows.Count - 1
# probably needs to add used.Column (so that it works if sheet has blank row/columns before data
col_count = used.Columns.Count
while last_cell is None:
    max_row_cell = used.Cells(row, col_count + 1).End(xldir.xlToLeft)
    max_row_col = max_row_cell.Column - 1
    # we need to check the first column is not filled
    if max_col > 0 or max_row_cell.Value != '':
        last_cell = (row, max_col)
    row -= 1
```

Reply:
How about something like this (untested!)?
```python
col_after_last = used.Column + used.Columns.Count + 1
for last_used_row in range(used.Row + used.Rows.Count, 0, -1):
    max_row_cell = used.Cells(last_used_row, col_after_last).End(xldir.xlToLeft)
    if max_row_cell.Column > 1 or max_row_cell.Value != '':
        break
row_after_last = last_used_row + 1
for last_used_col in range(used.Column + used.Columns.Count, 0, -1):
    max_col_cell = used.Cells(row_after_last, last_used_col).End(xldir.xlUp)
    if max_col_cell.Row > 1 or max_col_cell.Value != '':
        break
return (last_used_row, last_used_col)
```